Azure passthrough

joelddiaz opened the pull request:
Migrate away from depending on the Azure Active Directory Graph API since it will be sunset in June 2022.
- [x] No more mint mode for Azure. So continue supporting passthrough and manual mode as the only supported modes on Azure.
- [x] The implementation is to always mark the Azure root secret as passthrough, and pivot a live running originally-mint cluster to passthrough mode.
- [x] Overwrite existing Secrets with the content found in kube-system/azure-credentials (we can assume this as removing the root creds was never supported in Azure).
- [x] Try to clean up previously minted App Registrations / Service Principals (while the Azure AD Graph API still works), but treat errors as non-fatal (just set a condition on the CredentialsRequest).
- [x] Update test cases.

xref: https://issues.redhat.com/browse/CCO-173

akhil-rane commented on pkg/azure/actuator_test.go:
```diff
@@ -261,196 +269,281 @@ func TestActuator(t *testing.T) {
-				s := &corev1.Secret{}
-				// secret should be deleted
-				assert.Error(t, c.Get(context.TODO(),
-					types.NamespacedName{Name: cr.Spec.SecretRef.Name, Namespace: cr.Spec.SecretRef.Namespace},
-					s),
+
+				targetSecret := getCredRequestTargetSecret(t, c, cr)
+
+				rootSecret := getRootSecret(t, c)
+
+				// The targetSecret should be a copy of the root secret.
+				assertSecretEquality(t, rootSecret, targetSecret)
+			},
```
Do we need to check for `OrphanedCloudResource` condition here?

joelddiaz replied:
The condition is set by the credentials_request controller, so the best we can do with the test at the actuator-level is check for the right error returned by the actuator. (Although I should go ahead and add a test for the condition in the credentials_request controller. I'll work on that now).

joelddiaz commented on pkg/azure/actuator.go:
```diff
@@ -131,54 +121,43 @@ func (a *Actuator) needsUpdate(ctx context.Context, cr *minterv1.CredentialsRequ
-	)
-	if err != nil {
-		return true, fmt.Errorf("unable to create azure cred minter: %v", err)
+	if string(targetSecret.Data[AzureClientID]) != string(credentialsRootSecret.Data[AzureClientID]) ||
+		string(targetSecret.Data[AzureClientSecret]) != string(credentialsRootSecret.Data[AzureClientSecret]) {
+		return true, nil
 	}
```
It really shouldn't change as the cluster being pivoted from mint to passthrough will still be running under the same subscriptionID and tenantID. In fact, I might even argue that including subscription and tenant information as part of the azure secret format was an unfortunate mistake. That information should have been kept (and referenced when needed) in the Infrastructure CR since it is shared cluster-wide information.

joelddiaz commented on pkg/azure/actuator.go:
```diff
@@ -257,13 +204,18 @@ func (a *Actuator) Delete(ctx context.Context, cr *minterv1.CredentialsRequest)
 		string(credentialsRootSecret.Data[AzureSubscriptionID]),
 	)
 	if err != nil {
+		// TODO: Is it okay to treat this as an error worthy of retrying?
 		return fmt.Errorf("unable to create azure cred minter: %v", err)
```
Now with the passthrough steady-state, there really should never be any Azure API calls being made at all. So any kind of failure to make an Azure client is basically not a big deal.
But I agree, I don't really see why this should/would fail as there are no creds involved at that point.

joelddiaz commented on pkg/azure/actuator.go:
```diff
@@ -207,46 +186,14 @@ func (a *Actuator) Delete(ctx context.Context, cr *minterv1.CredentialsRequest)
 		return err
 	}
 
-	// When a service principal is deleted, it's corresponding credentials becomes invalid.
-	// Pass-through credentials are not created through crafted service principal.
-	// When a request is deleted, there is no service principal to delete.
-	// Thus, corresponding secret still provides valid credentials.
-	// For that reason, existing secret object needs to be deleted as well to avoid
-	// credentials leaking.
-	//
-	// Also, there is no harm in deleting the secret in general. Every component consuming
-	// the secret will be forbidden to talk to Azure API once the service principal is destroyed.
-	existingSecret := &corev1.Secret{}
-	err = a.client.Get(ctx, client.ObjectKey{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret)
-	if err == nil {
-		logger.Infof("Deleting secret %v/%v", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name)
-		if err := a.client.Delete(ctx, existingSecret); err != nil {
-			return fmt.Errorf("unable to delete secret %v/%v: %v", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name, err)
-		}
-	} else if !kerrors.IsNotFound(err) {
-		return fmt.Errorf("unable to get secret %v/%v: %v", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name, err)
-	}
-
-	if credentialsRootSecret.Annotations[constants.AnnotationKey] == constants.PassthroughAnnotation {
-		return nil
-	}
```
This code is already handled in the non-actuator-specific Delete() path https://github.com/openshift/cloud-credential-operator/blob/master/pkg/operator/credentialsrequest/credentialsrequest_controller.go#L436-L457
So this code was unnecessary to begin with.

joelddiaz commented on pkg/azure/actuator.go:
```diff
@@ -440,41 +323,52 @@ func (a *Actuator) updateProviderStatus(ctx context.Context, logger log.FieldLog
+		return cleanupErr
+	}
+	return syncErr
+}
```
This actually addresses the potential that we would fail to sync the secret, but still attempted a cleanup. That could lead to bad things. Good catch!

joelddiaz commented:
@akhil-rane i fixed up the code where you suggested and added a whole new set of test cases at the credentialsrequest_controller level to cover the new OrphanedCloudResource condition. PTAL

joelddiaz commented:
/test e2e-azure

joelddiaz commented:
/test verify

joelddiaz commented:
/test verify

joelddiaz commented:
/test e2e-azure-upgrade

joelddiaz commented:
/test e2e-azure-upgrade

joelddiaz commented:
/test verify

akhil-rane commented:
this looks good

joelddiaz commented:
@akhil-rane more changes to pull in an updated build-machinery-go package (that was failing the 'make verify' step).
Let me know when this looks okay, and I'll squash to make the PR look more reasonable.

joelddiaz commented:
/test e2e-azure-upgrade

akhil-rane commented:
yeah we can squash here, all good

joelddiaz commented:
/hold waiting for #426

joelddiaz commented:
/assign @jeana-redhat @lwan-wanglin @sferich888 
part 2 of removing support for Azure mint mode

joelddiaz commented:
/test e2e-azure

joelddiaz commented:
/test e2e-azure

joelddiaz commented:
/test e2e-azure

joelddiaz commented:
> @joelddiaz Here are the test cases for the card https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitems?query=trello%3ACCO%5C-173&sorting=id, PTAL! thanks. It's a little hard to reproduce the OrphanedCloudResource condition.
> 
Yes, this could be tricky. Perhaps you install an old version of OpenShift with Mint mode, disable CCO (either by marking it unmanaged in CVO and scaling down CVO, or move CCO to Manual mode), then swap the credentials in kube-system/azure-credentials with ones that can't do Mint mode, then upgrade to the new OpenShift/CCO, re-enable CCO, and it should migrate to Passthrough but fail to clean up the previously minted App Registrations.

> And here I have some questions:
> 
>     1. If we need to remove azure mint support from validation part on installer code ?https://github.com/openshift/installer/blob/02c23e64691b499d9cae9d534892e0424469b1b2/pkg/types/validation/installconfig.go#L610
>
Yes, good point. I'll add a Jira under the installer team.
 
>     2. If we need to doc things somewhere like "make sure .spec.credentialsMode of cloudcredential CR is not "Mint" before upgrade your cluster to 4.10" , otherwise the upgrade will fail, cco logs will show the below error
>
> 
> time="2021-12-17T07:31:13Z" level=error msg="error checking whether credentials already exists: invalid mode" controller=credreq cr=openshift-cloud-credential-operator/openshift-cloud-network-config-controller secret=openshift-cloud-network-config-controller/cloud-credentials
> time="2021-12-17T07:31:17Z" level=info msg="syncing credentials request" controller=credreq cr=openshift-cloud-credential-operator/azure-disk-csi-driver-operator

Yes, the docs will need some careful updates. cc @jeana-redhat

joelddiaz commented:
> @joelddiaz Here are the test cases for the card https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitems?query=trello%3ACCO%5C-173&sorting=id, PTAL! thanks. It's a little hard to reproduce the OrphanedCloudResource condition.
> 
Oh, and the test cases look good!

joelddiaz commented:
/test e2e-azure

sferich888 commented:
/label px-approved

joelddiaz commented:
@lwan-wanglin @jeana-redhat anything else needed before getting the qe/docs approve labels?
@akhil-rane anything else needed before an lgtm?

joelddiaz commented:
/test e2e-azure

akhil-rane commented:
/lgtm

joelddiaz commented:
/unhold
#246 has merged

lwan-wanglin commented:
@joelddiaz Thanks for your given steps , I can reproduce deleting SP failure case, and I filed a new  [test case](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitems?query=trello%3ACCO%5C-173) to cover it . I missed one check(verify SPs are deleted from Azure cloud) on previous test cases, now I update this step to the test cases, but I meet a situation, I find cco just remove role assignment, but not deleting applications, is it expected?  if not please let me know, I will file a bug to track.

For this PR, I think we can add qe-approved label.
/label qe-approved

jeana-redhat commented:
Will def be able to doc this, just need to know when each piece is landing. I know y'all will keep me updated on that :)
/label docs-approved

akhil-rane commented:
/cherry-pick release-4.9
